src/cue/contracts.cue

```diff
@@ -0,0 +1,76 @@
+package katapult
+
+contractBase: {
+  type?: string
+  slug?: string
+  data?: { ... }
+}
+
+ref :: contractBase & {
+  as?: string
+  cardinality?: string
+  version?: string
+}
+
+capabilityType: [Name=_]: contractBase & {
+  type: Name
+}
+
+capabilityType: endpoint: {
+  data: {
+    port: uint | *3000
+    protocol: "TCP" | "UDP" | *"TCP"
+  }
+}
+
+Contract :: contractBase & {
+  type: string
+  slug: string
+
+  requires?: [...ref]
+  provides?: [...ref]
+
+  // Default capabilities inferred from the type.
+  if (type == "sw.containerized-scalable-service" || type == "sw.containerized-service") {
+    provides: [...ref] | *[capabilityType.endpoint & { as: "main-endpoint" }]
+  }
+}
+
+contracts: [Name=_]: Contract & {
+  slug: Name
+}
+
+contracts: {
+  "balena-dashboard": {
+    type: "sw.containerized-scalable-service",
+    requires: [
+      { slug: "balena-api" },
+      { slug: "balena-data" }
```

Comment:
`contracts[$slug]` allows accessing a contract of a component by its slug - currently defined manually, ant it should be built as a result of blueprint interpretation and selection of specific versions.
Versions, in general, are not touched in this code.